src/main/java/net/snowflake/client/core/QueryContextCache.java

```diff
@@ -152,148 +137,163 @@ public void deserializeFromArrowBase64(String data) {
-   * @param priority the priority of the element
-   * @param opaqueContext the binary data of the opaque context
-   * @return a query context element
-   */
-  private static QueryContextElement createElement(
-      long id, long timestamp, long priority, byte[] opaqueContext) {
-    return new QueryContextElement(id, timestamp, priority, opaqueContext);
-  }
+private static QueryContextElement deserializeQueryContextElement(JsonNode node) throws IOException {
+  QueryContextElement entry = new QueryContextElement();
+    JsonNode idNode = node.path("id");
+    if (idNode.isNumber()) {
+        entry.setId(idNode.asLong());
+    }
 
-  /**
-   * Serialize a query context into a base64 encoded string using Arrow. This will be used to send
-   * the serialized query context to Snowflake clients.
-   *
-   * @return If succeeds, the base64 encoded representation of the query context. Otherwise, return
-   *     null.
-   */
-  public String serializeToArrowBase64() {
-    synchronized (this) {
-      logger.debug("serializeToArrowBase64() called");
-      // Log existing cache entries
-      logCacheEntries();
+    JsonNode timestampNode = node.path("timestamp");
+    if (timestampNode.isNumber()) {
+        entry.setReadTimestamp(timestampNode.asLong());
+    }
 
-      TreeSet<QueryContextElement> elements = getElements();
+    JsonNode priorityNode = node.path("priority");
+    if (priorityNode.isNumber()) {
+        entry.setPriority(priorityNode.asLong());
+    }
 
-      if (elements.size() == 0) return null;
-      try (VectorSchemaRoot root =
-          VectorSchemaRoot.create(QUERY_CONTEXT_SCHEMA, new RootAllocator())) {
-        root.setRowCount(elements.size());
-        int pos = 0;
-        for (final QueryContextElement elem : elements) {
-          serializeEntry(root, elem, pos++);
-        }
+    JsonNode contextNode = node.path("context");
+    if (contextNode.isTextual()) {
+        byte[] contextBytes = contextNode.asText().getBytes();
+        entry.setContext(contextBytes);
+    }
 
-        ByteArrayOutputStream out = new ByteArrayOutputStream();
-        try (ArrowStreamWriter writer = new ArrowStreamWriter(root, null, out)) {
-          writer.start();
-          writer.writeBatch();
-          writer.end();
-        }
+    return entry;
+}
+
+public String serializeQueryContextJson() {
+  synchronized (this) {
+    logger.debug("serializeQueryContextJson() called");
+    // Log existing cache entries
+    logCacheEntries();
 
-        String data = Base64.getEncoder().encodeToString(out.toByteArray());
+    TreeSet<QueryContextElement> elements = getElements();
 
-        logger.debug("serializeToArrowBase64(): data to send to server {}", data);
+    if (elements.size() == 0) return null;
 
-        return data;
-      } catch (Exception e) {
-        logger.debug("serializeToArrowBase64(): Exception {}", e.getMessage());
+    try{
+      int index = 0;
+      ObjectMapper mapper = new ObjectMapper();
```

Comment:
Don't allocate an object mapper here.

Reply:
I move the object mapper to a static variable and only initialize it for once.